### DISCUSSION TITLE
User subroutines for element, material, etc

### DISCUSSION THREAD

**Discussion by kengwit** [OpenRadioss/OpenRadioss#482] [Users]
Hi, first time looking at Altair/Radioss. Given that the source code is now accessible, is there any guidance or documentation on what would be the best way to add user material/element? Is there a "usermat" interface similar to LS-Dyna or does one have to add say material or element ID greater than XXX, and modify associated (driver) code accordingly?

**Comment by MarianBulla** [comment]
Hi @kengwit , yes Altari Radioss is supporting user subroutines (user materials, user elements, user failure criteria, ...) since many years.
One can create dynamic linked libraries with its own user laws and use it across the different Radioss versions.
You can find a detailed docu here: https://2022.help.altair.com/2022/hwsolvers/rad/topics/solvers/rad/introduction_user_code_r.htm
For OpenRadioss, a bug occured, which we are going to fix, now.
So, stay tuned, please. It will be also possible, soon.

**Reply by kengwit** [reply]
Hi @MarianBulla 

Thanks for the response. I looked through the document you mentioned. You mentioned that there is a "bug" - are you referring to the current lack of the "userlib_sdk" in OpenRadioss?

**Reply by MarianBulla** [reply]
Hi @kengwit 
Yes exactly. We are fixing it, so you will be able to use the USER_sdk also in OpenRadioss.
There are no issues in Altair Radioss (the commerzial version) because of to the compiler.
Will be fixed, soon. I will let you know.

**Comment by kengwit** [comment]
Thanks, looking forward to having this.

**Comment by MarianBulla** [comment]
Hi @kengwit,
the issue with the USER subroutines is fixed, now.
You can compile your USER materials and USER failure criteria and USER elements and USER .... and run it with OpenRadioss, now.

**Comment by kengwit** [comment]
Hi @MarianBulla 

Thank you. I will check these features out.

**Comment by kengwit** [comment]
@MarianBulla I am looking at the link you mentioned, specifically this one: 

https://2022.help.altair.com/2022/hwsolvers/rad/topics/solvers/rad/generate_user_libraries_user_code_r.htm#reference_dxp_sfc_j3b

I am looking at the linear elastic model described in PDF document [userscode_ch2_USER_Mat29_31.pdf](https://forum.altair.com/applications/core/interface/file/attachment.php?id=15160), mentioned in the link here:

https://community.altair.com/community?id=community_question&sys_id=37a600ba1b2bd0908017dc61ec4bcb91

and have the following questions:

1) Is there a "UserLib SDK"? For Windows, I am looking for a batch file "build_userlib_win64.bat" that compiles the "Starter source files" and "Engine Source files" to generate a "Library File".
2) In which folder do I put this "Library File"?

Thanks.

**Comment by kengwit** [comment]
@MarianBulla Any news on the availability of the UserLib SDK?

**Reply by MarianBulla** [reply]
Yes, it is available, now.
I'm still checking it.
But it looks very good. Works well.
We will do a bigger announcement, once we have put it on the right place.
Will come, soon.

**Reply by MarianBulla** [reply]
If you want, you can already try: https://github.com/OpenRadioss/userlib_sdk

**Reply by MarianBulla** [reply]
We have moved the userlib_sdk into the Tools repository, since we want to maintain such additional tools separately, with its own license, which makes it more flexible for all: https://github.com/OpenRadioss/Tools

**Reply by kengwit** [reply]
Thank you, I will check the Tools repo.

**Reply by MarianBulla** [reply]
Great! There is also an example, you can try. 
Let me know, if it works for you, please.
If you need other examples, stay tune, they will come...
Thanks and best regards,
Marian

**Comment by raas1saar-ux** [comment]
Hi Marian,

I am exploring a use case for implicit analysis for rubber materials and planning to make a hybrid user element for the analysis. Has anybody done this before? Is any document for reference or help is available?
Thanks

**Reply by MarianBulla** [reply]
Hi @raas1saar-ux,

thank you very much for your interest in OpenRadioss :-)

For rubber-kind materials, I would recommend you to use the deformation gradient tensor approach, which we have, if you use the total strain (I_ststr flag = 10 or 11 or 12). 
For the user materials, you have everything you need in the element buffer.

Let me know, if you need an example for a rubber material (like Mooney-Rivlin).

And for sure, you should be able to use it in implicite analysis, too. 

A new element type would be more that welcome for this kind of application. 

Let me know, if it works for you.

Thanks and best regards,
Marian

**Reply by raas1saar-ux** [reply]
> element buffer

Hi Marian,

Thanks for your message. Please let me know what is the element buffer you mention? Also, please share any examples of rubber material, like Mooney-Rivlin, Ogden, etc. 

Regards,
Kartik

**Reply by MarianBulla** [reply]
Hi Kartik,
yes sure. Feel free to test. 
Here is the source code and an example (for explicite simulation). 
So, you need to modify the model for implicite.
Let me know, how it running on your site.

Thanks and best,
[Mooney_Rivlin.zip](https://github.com/user-attachments/files/24185308/Mooney_Rivlin.zip)

Marian

**Comment by raas1saar-ux** [comment]
Thanks Marian. I will update you.

Regards,
Kartik
